app/UI/app_update.tsx

```diff
@@ -277,10 +280,79 @@ const handle_update_event = async (chan: PutChannel<true>, args: {
                     chan.put(true);
                 }
             });
-        } else if (event.type == "UpdateMessageFiles") {
-            console.log("to be implemented");
-        } else if (event.type == "UpdateEditorText") {
-            console.log("to be implemented");
+        } else if (event.type == "UploadImages") {
+            function validateImageFile(file: File) {
+                // https://nostr.build/.well-known/nostr/nip96.json
+                const max_byte_size = 10485760;
+                const image_postfix = [".png", ".jpg", ".jpeg", ".gif", ".webp"];
+                if (file.size > max_byte_size) {
+                    return false;
+                }
+                if (!image_postfix.some((postfix) => file.name.endsWith(postfix))) {
+                    return false;
+                }
+                return true;
+            }
+            async function uploadValidFiles(app: any, validFiles: File[], apiURL: string): Promise<{
+                uploadedURLs: string[];
+                errorMessages: string[];
+            }> {
+                const uploadedURLs: string[] = [];
+                const errorMessages: string[] = [];
+                const uploadPromises = validFiles.map(async (file) => {
+                    try {
+                        const uploaded = await uploadFile(app.ctx, { api_url: apiURL, file });
+
+                        if (uploaded instanceof Error) {
+                            errorMessages.push(`${file.name}: ${uploaded.message}`);
+                        } else if (uploaded.status === "error") {
+                            errorMessages.push(`${file.name}: ${uploaded.message}`);
+                        } else {
+                            uploadedURLs.push(uploaded.nip94_event.tags[0][1]);
+                        }
+                    } catch (error) {
+                        errorMessages.push(`${file.name}: ${error.message}`);
+                    }
+                });
+                await Promise.all(uploadPromises);
+                return { uploadedURLs, errorMessages };
+            }
+            const { files } = event;
+            if (files.length === 0) {
+                app.toastInputChan.put(() => "no files selected");
+                continue;
+            }
+            const invalid_files_name: string[] = [];
+            const valid_files: File[] = [];
+            files.forEach((file) => {
```

Comment:
Please, don't default to `forEach`. This has been stated many times.

Reply:
Thanks, I'll avoid it later.